### PR TITLE
Fix #10534 - Chiller:Absorption does not autosize in some cases

### LINKED ISSUE #10534: Chiller:Absorption does not autosize in some cases
Issue overview
--------------
User file with Chiler:Absorption is auotsizing to zero. The problem is here:
https://github.com/NREL/EnergyPlus/blob/1b333b500c96960b1a1978657b9a6a4a28c9dd72/src/EnergyPlus/ChillerAbsorption.cc#L954
and here:
https://github.com/NREL/EnergyPlus/blob/1b333b500c96960b1a1978657b9a6a4a28c9dd72/src/EnergyPlus/ChillerAbsorption.cc#L1060

These should both be comparing to `HVAC::SmallWaterVolFlow`.

This got broken over 4 years ago and slipped through because the one example file with this chiller type is not autosized. And even if it was, if the design flow rate was large enough it would have made it through.

For this particular file, changing to Timestep,60; gets past these ifs and the chiller will autosize properly.

### Details
Some additional details for this issue (if relevant):
 - Version of EnergyPlus 24.1
 - Helpdesk ticket number 16669

### Checklist
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [x] Defect file added EnergyPlusDevSupport\DefectFiles\...
 - [ ] Ticket added to Pivotal for defect (development team task)
 - [ ] Pull request created (the pull request will have additional tasks related to reviewing changes that fix this defect)

### PR DESCRIPTION
**Pull request by jmarrec**
Pull request overview
---------------------

 - Fixes #10534

### Pull Request Author
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [x] Title of PR should be user-synopsis style (clearly understandable in a standalone changelog context)
 - [x] Label the PR with at least one of: Defect, Refactoring, NewFeature, Performance, and/or DoNoPublish
 - [x] Pull requests that impact EnergyPlus code must also include unit tests to cover enhancement or defect repair
 - [x] Author should provide a "walkthrough" of relevant code changes using a GitHub code review comment process
 - [x] If any diffs are expected, author must demonstrate they are justified using plots and descriptions
 - [x] If changes fix a defect, the fix should be demonstrated in plots and descriptions
 - [x] If any defect files are updated to a more recent version, upload new versions here or on DevSupport
 - [x] If IDD requires transition, transition source, rules, ExpandObjects, and IDFs must be updated, and add IDDChange label
 - [x] If structural output changes, add to output rules file and add OutputChange label
 - [x] If adding/removing any LaTeX docs or figures, update that document's CMakeLists file dependencies

### Reviewer
This will not be exhaustively relevant to every PR.
 - [ ] Perform a Code Review on GitHub
 - [ ] If branch is behind develop, merge develop and build locally to check for side effects of the merge
 - [ ] If defect, verify by running develop branch and reproducing defect, then running PR and reproducing fix
 - [ ] If feature, test running new feature, try creative ways to break it
 - [ ] CI status: all green or justified
 - [ ] Check that performance is not impacted (CI Linux results include performance check)
 - [ ] Run Unit Test(s) locally
 - [ ] Check any new function arguments for performance impacts
 - [ ] Verify IDF naming conventions and styles, memos and notes and defaults
 - [ ] If new idf included, locally check the err file and other outputs

### CODE REVIEW

File: tst/EnergyPlus/unit/ChillerAbsorption.unit.cc [hunk 1]
```diff
@@ -1830,3 +1834,265 @@ TEST_F(EnergyPlusFixture, ChillerAbsorption_Calc)
...
+
+    EXPECT_TRUE(compare_enums(thisChiller.FlowMode, DataPlant::FlowMode::Constant));
+    EXPECT_TRUE(thisChiller.GenInputOutputNodesUsed);
+    EXPECT_GT(thisChiller.GeneratorInletNodeNum, 0);
+    EXPECT_GT(thisChiller.GeneratorOutletNodeNum, 0);
+    EXPECT_TRUE(compare_enums(thisChiller.GenHeatSourceType, DataLoopNode::NodeFluidType::Water));
+    EXPECT_TRUE(thisChiller.GeneratorVolFlowRateWasAutoSized);
+
+    EXPECT_DOUBLE_EQ(2.0, thisChiller.GeneratorSubcool);
+    EXPECT_DOUBLE_EQ(1.5, thisChiller.SizFac);
```
Comment (jmarrec) [inline tst/EnergyPlus/unit/ChillerAbsorption.unit.cc:1941]: New test. Start by validating the GetInput (`ChillerAbsorption::GetBLASTAbsorberInput`) routine, make sure it reads the right fields.

File: tst/EnergyPlus/unit/ChillerAbsorption.unit.cc [hunk 2]
```diff
@@ -1830,3 +1834,265 @@ TEST_F(EnergyPlusFixture, ChillerAbsorption_Calc)
...
+
+        thisLoopSizing.PlantLoopName = thisLoop.Name;
+        thisLoopSizing.LoopType = DataSizing::TypeOfPlantLoop::Heating;
+        thisLoopSizing.ExitTemp = 80.0;
+        thisLoopSizing.DesVolFlowRate = 0.01;
+        thisLoopSizing.DeltaT = 20.0;
+    }
+
+    state->dataPlnt->PlantFirstSizesOkayToFinalize = true;
+    state->dataPlnt->PlantFinalSizesOkayToReport = true;
```
Comment (jmarrec) [inline tst/EnergyPlus/unit/ChillerAbsorption.unit.cc:2022]: Set up plant loop and Sizing:Plant objects for all three loops.

File: tst/EnergyPlus/unit/ChillerAbsorption.unit.cc [hunk 3]
```diff
@@ -1830,3 +1834,265 @@ TEST_F(EnergyPlusFixture, ChillerAbsorption_Calc)
...
+                                                                 state->dataSize->PlantSizData(genLoopNum).ExitTemp,
+                                                                 state->dataPlnt->PlantLoop(genLoopNum).FluidIndex,
+                                                                 "ChillerAbsorption_Autosize_TEST");
+
+    Real64 const rho_gen = FluidProperties::GetDensityGlycol(*state,
+                                                             state->dataPlnt->PlantLoop(genLoopNum).FluidName,
+                                                             (state->dataSize->PlantSizData(genLoopNum).ExitTemp - SteamDeltaT),
+                                                             state->dataPlnt->PlantLoop(genLoopNum).FluidIndex,
+                                                             "ChillerAbsorption_Autosize_TEST");
+    Real64 const expectedGeneratorVolFlowRate = (expectedNomCap * SteamInputRatNom) / (Cp_gen * rho_gen * SteamDeltaT);
```
Comment (jmarrec) [inline tst/EnergyPlus/unit/ChillerAbsorption.unit.cc:2072]: We'll be validating the sizing results, so pre-calc them.

File: tst/EnergyPlus/unit/ChillerAbsorption.unit.cc [hunk 4]
```diff
@@ -1830,3 +1834,265 @@ TEST_F(EnergyPlusFixture, ChillerAbsorption_Calc)
...
+    EXPECT_NE(thisChiller.NomPumpPower, DataSizing::AutoSize);
+    EXPECT_NE(thisChiller.EvapVolFlowRate, DataSizing::AutoSize);
+    EXPECT_NE(thisChiller.CondVolFlowRate, DataSizing::AutoSize);
+    EXPECT_NE(thisChiller.GeneratorVolFlowRate, DataSizing::AutoSize);
+
+    EXPECT_GT(thisChiller.NomCap, 0.0);
+    EXPECT_GT(thisChiller.NomPumpPower, 0.0);
+    EXPECT_GT(thisChiller.EvapVolFlowRate, 0.0);
+    EXPECT_GT(thisChiller.CondVolFlowRate, 0.0);
+    EXPECT_GT(thisChiller.GeneratorVolFlowRate, 0.0);
```
Comment (jmarrec) [inline tst/EnergyPlus/unit/ChillerAbsorption.unit.cc:2091]: Size. Ensure it's not autosized and not zero. This is enough to demonstrate #10534 occurs.

Cf commit message in f7967682aa970063d9a0bf40c1c6fc1a87b1ca20

```
Expected equality of these values:
  expectedNomCap
    Which is: 314812.63583550003
  thisChiller.NomCap
    Which is: 0
/home/julien/Software/Others/EnergyPlus/tst/EnergyPlus/unit/ChillerAbsorption.unit.cc:2094: Failure
Expected equality of these values:
  expectedNomCap * nomCapToPumpRatio
    Which is: 1416.65686125975
  thisChiller.NomPumpPower
    Which is: 0
/home/julien/Software/Others/EnergyPlus/tst/EnergyPlus/unit/ChillerAbsorption.unit.cc:2095: Failure
Expected equality of these values:
  expectedEvapVolFlowRate
    Which is: 0.014999999999999999
  thisChiller.EvapVolFlowRate
    Which is: 0
/home/julien/Software/Others/EnergyPlus/tst/EnergyPlus/unit/ChillerAbsorption.unit.cc:2096: Failure
Expected equality of these values:
  expectedCondVolFlowRate
    Which is: 0.030196473750000008
  thisChiller.CondVolFlowRate
    Which is: 0
/home/julien/Software/Others/EnergyPlus/tst/EnergyPlus/unit/ChillerAbsorption.unit.cc:2097: Failure
Expected equality of these values:
  expectedGeneratorVolFlowRate
    Which is: 0.0038118078460457113
  thisChiller.GeneratorVolFlowRate
    Which is: 0
```

File: tst/EnergyPlus/unit/ChillerAbsorption.unit.cc [hunk 5]
```diff
@@ -1830,3 +1834,265 @@ TEST_F(EnergyPlusFixture, ChillerAbsorption_Calc)
...
+
+    EXPECT_DOUBLE_EQ(expectedNomCap, thisChiller.NomCap);
+    EXPECT_DOUBLE_EQ(expectedNomCap * nomCapToPumpRatio, thisChiller.NomPumpPower);
+    EXPECT_DOUBLE_EQ(expectedEvapVolFlowRate, thisChiller.EvapVolFlowRate);
+    EXPECT_DOUBLE_EQ(expectedCondVolFlowRate, thisChiller.CondVolFlowRate);
+    EXPECT_DOUBLE_EQ(expectedGeneratorVolFlowRate, thisChiller.GeneratorVolFlowRate);
```
Comment (jmarrec) [inline tst/EnergyPlus/unit/ChillerAbsorption.unit.cc:2097]: Ensure calculation is what we expect.

File: src/EnergyPlus/ChillerAbsorption.cc [hunk 6]
```diff
@@ -951,7 +951,7 @@ void BLASTAbsorberSpecs::sizeChiller(EnergyPlusData &state)
     }
 
     if (PltSizNum > 0) {
-        if (state.dataSize->PlantSizData(PltSizNum).DesVolFlowRate >= state.dataHVACGlobal->TimeStepSys) {
+        if (state.dataSize->PlantSizData(PltSizNum).DesVolFlowRate >= HVAC::SmallWaterVolFlow) {
```
Comment (jmarrec) [inline src/EnergyPlus/ChillerAbsorption.cc:954]: Fix that @mjwitte had flagged here

File: src/EnergyPlus/ChillerAbsorption.cc [hunk 7]
```diff
@@ -1057,7 +1057,7 @@ void BLASTAbsorberSpecs::sizeChiller(EnergyPlusData &state)
...
```
Comment (jmarrec) [inline src/EnergyPlus/ChillerAbsorption.cc:1060]: and here

File: src/EnergyPlus/ChillerAbsorption.cc [hunk 8]
```diff
@@ -1118,7 +1118,7 @@ void BLASTAbsorberSpecs::sizeChiller(EnergyPlusData &state)
     PlantUtilities::RegisterPlantCompDesignFlow(state, this->EvapInletNodeNum, tmpEvapVolFlowRate);
 
     if (PltSizCondNum > 0 && PltSizNum > 0) {
-        if (this->EvapVolFlowRate >= state.dataHVACGlobal->TimeStepSys && tmpNomCap > 0.0) {
+        if (this->EvapVolFlowRate >= HVAC::SmallWaterVolFlow && tmpNomCap > 0.0) {
```
Comment (jmarrec) [inline src/EnergyPlus/ChillerAbsorption.cc:1121]: Found this one thanks to unit test.

File: src/EnergyPlus/ChillerAbsorption.cc [hunk 9]
```diff
@@ -1197,7 +1197,7 @@ void BLASTAbsorberSpecs::sizeChiller(EnergyPlusData &state)
 
     if ((PltSizSteamNum > 0 && this->GenHeatSourceType == DataLoopNode::NodeFluidType::Steam) ||
         (PltSizHeatingNum > 0 && this->GenHeatSourceType == DataLoopNode::NodeFluidType::Water)) {
-        if (this->EvapVolFlowRate >= state.dataHVACGlobal->TimeStepSys && tmpNomCap > 0.0) {
+        if (this->EvapVolFlowRate >= HVAC::SmallWaterVolFlow && tmpNomCap > 0.0) {
```
Comment (jmarrec) [inline src/EnergyPlus/ChillerAbsorption.cc:1200]: And this one as well.

**Comment by jmarrec** [comment]
Taking the defect file, the sizing is no longer 0 with the fix (simulation fails to run because of plant temperatures, but that's not the point of this PR).

EnergyPlusDevSupport/DefectFiles/10000s/10534

image

File: src/EnergyPlus/ChillerAbsorption.cc (hunk 6 above)
Comment (Myoldmopar) [reply src/EnergyPlus/ChillerAbsorption.cc:954]: lololol vol flow > time step?  Maybe we should convert Real64 to a units-backed floating point type 

JUST KIDDING EVERYONE!

File: src/EnergyPlus/ChillerAbsorption.cc (hunk 8 above)
Comment (Myoldmopar) [reply src/EnergyPlus/ChillerAbsorption.cc:1121]: oh my gosh.  OK, getting more serious about a units backed floating point representation.

STILL KIDDING ... for now.

File: src/EnergyPlus/ChillerAbsorption.cc (hunk 9 above)
Comment (Myoldmopar) [reply src/EnergyPlus/ChillerAbsorption.cc:1200]: BARELY kidding about it at this point.

**Review by Myoldmopar — APPROVED** [review]
Excellent catch @jmarrec and @mjwitte.  No problem with this, clear fixes and clear unit testing with demonstration of failed test.

**Comment by Myoldmopar** [comment]
Tests happily here, let's get this merged.  Thanks @jmarrec and @mjwitte

**Pull request merged**